fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -1363,21 +1367,154 @@ Reference<LogGenerationData> findLogData(
+	// this might be an off-by-one error from other places.
+	if (req.beginVersion <= logData->persistentDataDurableVersion) {
+		// Just in case the durable version changes while we are waiting for the read, we grab this data from memory. We
+		// may or may not actually send it depending on whether we get enough data from disk. SOMEDAY: Only do this if
+		// an initial attempt to read from disk results in insufficient data and the required data is no longer in
+		// memory SOMEDAY: Should we only send part of the messages we collected, to actually limit the size of the
+		// result?
+
+		if (reqOnlySpilled) {
+			endVersion = logData->persistentDataDurableVersion + 1;
+		} else {
+			peekMessagesFromMemory(req, values, beginVersion, endVersion, logData);
+		}
+
+		if (logData->shouldSpillByValue(req.storageTeamID)) {
+			RangeResult kvs = wait(tlogGroup->second->persistentData->readRange(
+			    KeyRangeRef(persistStorageTeamMessagesKey(logData->logId, req.storageTeamID, req.beginVersion),
+			                persistStorageTeamMessagesKey(
+			                    logData->logId, req.storageTeamID, logData->persistentDataDurableVersion + 1)),
+			    TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA,
+			    TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA));
+			bool first = true;
+			for (auto& kv : kvs) {
+				auto ver = decodeStorageTeamMessagesKey(kv.key);
+				// try decode kv.value here, who is encoded by proxy
+				versionsFromDisk.insert(ver);
+				serializer.writeSerializedVersionSection(
+				    kv.value); // if comment this line, then can read the in-memory data, so problem is here
+				if (first) {
+					beginVersion = ver;
+					first = false;
+				}
+			}
+			if (kvs.expectedSize() >=
+			    TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA) { // if enough from disk, not reading from memory
+				endVersion = decodeStorageTeamMessagesKey(kvs.end()[-1].key) + 1;
+				onlySpilled = true;
+			} else {
+				serializeMemoryData(values, serializer, versionsFromDisk);
+			}
+		} else {
+			// TODO: make it work for spill by reference,
+			// now writing(and only writing) those in-memory data without considering on-disk data
+			serializeMemoryData(values, serializer);
+		}
+	} else {
+		if (reqOnlySpilled) {
+			endVersion = logData->persistentDataDurableVersion + 1;
+		} else {
+			peekMessagesFromMemory(req, values, beginVersion, endVersion, logData);
+			serializeMemoryData(values, serializer);
+		}
+		//TraceEvent("TLogPeekResults", self->dbgid).detail("ForAddress", replyPromise.getEndpoint().getPrimaryAddress()).detail("MessageBytes", messages.getLength()).detail("NextEpoch", next_pos.epoch).detail("NextSeq", next_pos.sequence).detail("NowSeq", self->sequence.getNextSequence());
+	}
+
+	TLogPeekReply reply;
+	reply.maxKnownVersion = logData->version.get();
+	reply.minKnownCommittedVersion = logData->minKnownCommittedVersion;
+	reply.data = StringRef(reply.arena, serializer.getSerialized());
+	reply.endVersion = endVersion;
+	reply.beginVersion = beginVersion;
```

Comment:
What is `beginVersion` used for? In what case, it is different from req.beginVersion?

Reply:
yeah, req.beginVersion is a begin of a range read, but this is the real begin version, I am not sure how it would be used though, as I am not super familiar with SS code, but code was there in `TLogPeekReply` and I thought there is a reason that it got merged that way, so I just make it work.